Create recipe

fmolinalopez opened the pull request:
Faltan test por falta de tiempo

Closes #22

fmolinalopez committed:
Creada funcion createRecipe.js

fmolinalopez committed:
Copiado package.json de master

Juanlu1995 commented on src/js/pages/edit/createRecipe.js:
```diff
@@ -0,0 +1,56 @@
+		try {
+			const recipe = new Recipe(recipeId, name, ingredients, steps, typeOfFood, origin, authorId);
+			success = true;
+			addRecipeToState(recipe);
```
deberías invertir el orden: primero añadirla, y si está bien, success = true

Juanlu1995 commented on src/js/pages/edit/createRecipe.js:
```diff
@@ -0,0 +1,56 @@
+			addRecipeToState(recipe);
+		} catch (err) {
+			success = false;
+			throw error = new Error(invalidRecipe);
```
creo que no es la forma correcta de lanzar excepciones try...catch:

try {
    // puede lanzar tres tipos de excepciones
    myroutine();  
} catch (e) {
    if (e instanceof TypeError) {
        // instrucciones para manejar excepciones TypeError
    } else if (e instanceof RangeError) {
        // instrucciones para manejar excepciones RangeError
    } else if (e instanceof EvalError) {
        // instrucciones para manejar excepciones EvalError
    } else {
       // instrucciones para manejar cualquier excepción no especificada
       logMyErrors(e); // pasa el objeto de la excepción al manejador de errores
}

MeLlamoPablo replied:
En este caso tendría más sentido lanzar nuestra propia instancia de un error personalizado que herede de `Error`, y si no es esa instancia, entonces es un error distinto. Pero me parece overkill para lo que estamos haciendo.

Juanlu1995 commented on src/js/pages/edit/createRecipe.js:
```diff
@@ -0,0 +1,56 @@
+		success = false;
+		throw error = new Error(repeatedIngredient);
+	}
+	return { success, error };
```
si metes el return en un finally queda más elegante y creo que también más correcto

Juanlu1995 reviewed (commented):
Los comentarios que te he hecho se basan en mi poca experiencia, hasta que un jedi no los corrija no los tomes por algo final, si no más bien como una opinion (por eso los "creo")

MeLlamoPablo commented on src/js/pages/edit/createRecipe.js:
```diff
@@ -0,0 +1,56 @@
+		/* Si la receta recibe valores invalidos lanzara un error, sino la añade
+		 a las recetas de state */
+		try {
+			const recipe = new Recipe(recipeId, name, ingredients, steps, typeOfFood, origin, authorId);
```
Aquí sería importante una de dos:

* Validar que no exista una receta con el ID que ha pasado el usuario, o mejor aún...
* No pedir el ID al usuario y asignarlo automáticamente

Ten en cuenta que eso no se valida en el constructor de Recipe porque éste es agnóstico del estado de la aplicación.

MeLlamoPablo commented on src/js/pages/edit/createRecipe.js:
```diff
@@ -0,0 +1,56 @@
+			addRecipeToState(recipe);
+		} catch (err) {
+			success = false;
+			throw error = new Error(invalidRecipe);
```
¡Esta función no debe lanzar excepciones! Debe devolver valores. Si se produce una excecpión, no la vuelvas a lanzar. Devuelve un `{ success: false, errorMessage: error.message }`.

fmolinalopez replied:
Con los cambios que le he hecho solo lanza excepciones la receta si le pasas datos insuficientes, no se como hacer que no las lanze

MeLlamoPablo committed:
Creada test suite vacía para createRecipe

MeLlamoPablo committed:
Merge branch 'master' into createRecipe

fmolinalopez committed:
Ya no deberia lanzar excepciones

fmolinalopez committed:
Merge branch 'master' of https://github.com/adaits-2daw-cliente-grupo1/ejercicio-1 into createRecipe

fmolinalopez committed:
Merge branch 'createRecipe' of https://github.com/adaits-2daw-cliente-grupo1/ejercicio-1 into createRecipe

MeLlamoPablo committed:
Cambios finales

- La función ya no recibe el parámetro ID sino que lo calcula
- Implementados los tests
